src/domain/waste-balances/calculator.js

```diff
@@ -119,27 +119,75 @@ export const calculateWasteBalanceUpdates = ({
   let currentAmount = currentBalance.amount || 0
   let currentAvailableAmount = currentBalance.availableAmount || 0
 
+  // Optimization: Pre-calculate credited amounts to avoid O(N*M) complexity
+  const creditedAmountMap = new Map()
+
+  /**
+   * Updates the credited amount map with the transaction amount for each entity.
+   * Credits increase the credited amount, Debits decrease it.
+   * @param {import('#domain/waste-balances/model.js').WasteBalanceTransaction} transaction
+   */
+  const updateCreditedAmountMap = (transaction) => {
+    const isCredit = transaction.type === WASTE_BALANCE_TRANSACTION_TYPE.CREDIT
+    const isDebit = transaction.type === WASTE_BALANCE_TRANSACTION_TYPE.DEBIT
+
+    if (!isCredit && !isDebit) {
+      return
+    }
+
+    const sign = isCredit ? 1 : -1
+    const netAmount = transaction.amount * sign
```

Comment:
Should we use something like decimal.js to make sure we don't get any creative calculations from JS?